concent_api/core/payments/storage.py

```diff
@@ -0,0 +1,145 @@
+import logging
+from typing import Dict
+from typing import List
+
+from ethereum.transactions import Transaction
+from golem_messages.utils import encode_hex
+#from golem_sci.transactionsstorage import TransactionsStorage
+
+from django.db import transaction
+
+from core.models import GlobalTransactionState
+from core.models import PendingEthereumTransaction
+from core.payments import service as payments_service
+
+
+logger = logging.getLogger(__name__)
+
+
+class DatabaseTransactionsStorage():
+    """
+    Concent custom implementation of TransactionsStorage interface used to store Ethereum transaction data into
+    database using Django models.
+    """
+
+    def __init__(self, *args: List, **kwargs: Dict) -> None:
+        super().__init__(*args, **kwargs)
+
+        if not GlobalTransactionState.objects.filter(pk=0).exists():
+            global_transaction_state = GlobalTransactionState(
+                pk=0,
+                nonce=payments_service.get_transaction_count(),  # pylint: disable=no-value-for-parameter
+            )
+            global_transaction_state.full_clean()
+            global_transaction_state.save()
+
+    @transaction.atomic(using='control')
+    def get_nonce(self) -> int:
+        """
+        Returns the nonce for the next transaction.
+        """
+        global_transaction_state = self._get_locked_global_transaction_state()
+        return global_transaction_state.nonce
+
+    @transaction.atomic(using='control')  # pylint: disable=no-self-use
+    def get_all_tx(self) -> List[Transaction]:
+        """
+        Returns the list of all transactions.
+        """
+        return [
+            Transaction(
+                nonce    = int(ethereum_transaction.nonce),
+                gasprice = int(ethereum_transaction.gasprice),
+                startgas = int(ethereum_transaction.startgas),
+                value    = int(ethereum_transaction.value),
+                v        = int(ethereum_transaction.v),
+                r        = int(ethereum_transaction.r),
+                s        = int(ethereum_transaction.s),
+                to       = encode_hex(ethereum_transaction.to.tobytes()),
+                data     = ethereum_transaction.data.tobytes(),
+            )
+            for ethereum_transaction in PendingEthereumTransaction.objects.all()
+        ]
+
+    @transaction.atomic(using='control')
+    def put_tx_and_inc_nonce(self, tx: Transaction) -> None:
+        """
+        Save a valid transaction and increase the nonce.
+        """
+        global_transaction_state = self._get_locked_global_transaction_state()
+
+        if global_transaction_state.nonce != tx.nonce:
+            raise Exception(
+                f'Transaction nonce does not match. '
+                f'Current={global_transaction_state.nonce}, '
+                f'Transaction={tx.nonce}'
```

Comment:
I have followed up the fact that `JsonTransactionsStorage` is raising pure `Exception`, and assumed that it might matter, but if you think it doesn't then off course, lets make it a custom exception class.

Reply:
Let's use a custom class. Even if it was checked in any way, a custom exception would inherit from `Exception` anyway.

 We actually should add it to our coding style: https://github.com/Code-Poets/coding-style/issues/10